### ISSUE TITLE
List of editor integration guides

### ISSUE THREAD

**Issue by chrisdone** [haskell-lang/haskell-lang#21]
In the documentation section, let's have an editor integration section.

We can list the common ways to get editor integration, and consult the relevant people for the proper way to do it for a given environment.
- Emacs
- Spacemacs
- Vim
- Sublime
- Atom
- etc

**Comment by bitemyapp** [comment]
Going to go ahead and put this stake in the ground: don't recommend ghc-mod. It's extremely fragile, wasted a lot of time for beginners, and most would be fine with just the errors from flycheck + Stack (for Emacs anyway. There's an equiv for Vim) and a REPL for the rest.

I have a basic but reliable and useful config in my dotfiles at: https://github.com/bitemyapp/dotfiles/blob/master/.emacs.d/haskell/haskell-config.el

I could substantially reduce into a complete, working, but minimal example if the site can make use of it.

**Comment by dpwiz** [comment]
> Going to go ahead and put this stake in the ground: don't recommend ghc-mod. It's extremely fragile, wasted a lot of time for beginners

Sad, but true.

**Comment by chrisdone** [comment]
Agreed. I made flycheck-stack because I couldn't get ghc-mod to work.

**Comment by vincenthz** [comment]
for atom, `ghc-mod` seems to work reasonably and it's pretty much the only way to have advanced IDE features (?);

**Comment by rvion** [comment]
@vincenthz ghc-mod works great but is really slow for large projects (much slower than ghci-based tools) Sadly, since ghci, intero, ghcid, ghci-ng , etc don't implement a proper "ide-server" (with proper api), it's a bit harder to develop ide plugins that use them because you have to manage the low level process interraction yourself.

**Comment by rvion** [comment]
for atom, hdevtools is also easy to setup, but only gives errors, and not much more. the type tooltip  provided by its companion plugin is not great.
for sublime text, sublimehaskell also works decently, but is quite slow as it is based on ghc-mod.

If you only want error messages + notifications in the gutter, you can also use `ghcid -o errors-and-warnings.txt` along a custom script to pares and print errors and warnings in the errors-and-warnings.txt dump file. works to some extent, but not great

**Comment by bitemyapp** [comment]
My only hesitation here is that the IDE tooling story is fairly immature and I don't want to give new people the impression things "just work" here — they do not outside of perhaps `ghcid` and `intero`. I'm a bad test case for Intero as my work project forces me to restart it manually several times an hour. (I modify dependencies a lot) I think it's quite good for ordinary users, albeit only available for Emacs at the moment.

Looks like there's been some push to make intero work on neovim: https://www.reddit.com/r/haskell/comments/4n27r2/intero_for_neovim/

**Comment by chrisdone** [comment]
I also restart intero fairly often for adding dependencies. I'll add some automation around that eventually.

**Comment by bitemyapp** [comment]
@chrisdone in my case, it's modifying deps rather than adding, but even covering new deps would be lovely. Most folks are more likely to be affected by addition rather than modification as well. Automation for either/or would be great :D
